Rewrite the `ci.py` script in Rust

Kobzol opened the pull request:
It would seem that I would learn by now that any script written in Python will become unmaintainable sooner or later, but alas..

r? @marcoieni

try-job: aarch64-gnu
try-job: dist-x86_64-linux-alt
try-job: x86_64-msvc-ext2

Fixes: https://github.com/rust-lang/rust/issues/137013

Kobzol committed:
Implement CI tooling in Rust

Kobzol committed:
Add missing base-job directive

Kobzol committed:
Add local job execution to `citool`

Kobzol committed:
Use citool in CI

Kobzol committed:
Delete the previous Python script

Kobzol committed:
Update documentation

jieyouxu commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,375 @@
+const DOCKER_DIRECTORY: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/../docker");
+const JOBS_YML_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/../github-actions/jobs.yml");
+
+/// Representation of a job loaded from the jobs.yml file.
```
```suggestion
/// Representation of a job loaded from the `src/ci/github-actions/jobs.yml` file.
```

jieyouxu commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,375 @@
+struct Job {
+    /// Name of the job, e.g. mingw-check
+    name: String,
+    /// GitHub runner on which the job should be executed
+    os: String,
```
Nit: maybe `runner_name`?

jieyouxu commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,375 @@
+    let db = std::fs::read_to_string(path)?;
+    let mut db: Value = serde_yaml::from_str(&db)?;
+
+    // We need to expand merge keys (<<), because serde_yaml can't deal with them
+    // `apply_merge` only applies the merge once, so do it a few times to unwrap nested merges.
+    db.apply_merge()?;
+    db.apply_merge()?;
```
Question: does `serde_yaml` have a way to `apply_merge`-to-fixpoint? I suppose this is is fine though.

jieyouxu replied:
Yeah I don't think this is needed. If anything, I feel like having more than 2 layers of nesting is... suspicious.

jieyouxu commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,375 @@
+                }
+                if !unknown_jobs.is_empty() {
+                    return Err(anyhow::anyhow!(
+                        "Custom job(s) `{}` not found in auto jobs",
+                        unknown_jobs.join(", ")
```
Discussion: do we want to list available auto jobs? Though prob. overkill.

Kobzol replied:
With tiny exceptions, I mostly wanted to port the Python logic 1:1, to avoid doing too much in this PR. Further improvements are welcome in follow-up PRs :)

jieyouxu commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,375 @@
+/// Representation of a job outputted to a GitHub Actions workflow.
+#[derive(serde::Serialize, Debug)]
+struct GithubActionsJob {
+    name: String,
+    full_name: String,
```
Question: what's the distinction between `name` and `full_name`? Maybe good as a comment?

Kobzol replied:
Will add comments, thanks.

jieyouxu commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,375 @@
+        RunType::AutoJob => "auto",
+    };
+
+    eprintln!("Output:\njobs={jobs:?}\nrun_type={run_type}");
```
Nit (style):

```suggestion
    eprintln!("Output");
    eprintln!("jobs={jobs:?}");
    eprintln!("run_type={run_type}");
```

jieyouxu reviewed (commented):
Looks very cool. Not an infra reviewer, just some drive-by comments :P

marcoieni commented on src/ci/github-actions/jobs.yml:
```diff
@@ -51,6 +51,7 @@ runners:
     # Free some disk space to avoid running out of space during the build.
     free_disk: true
     os: ubuntu-22.04-arm
+    <<: *base-job
```
this is missing in job-aarch64-linux-8c too

Kobzol replied:
Interesting that it didn't crash. Thanks!

Kobzol committed:
Apply review comments

marcoieni commented on src/doc/rustc-dev-guide/src/building/optimized-build.md:
```diff
@@ -126,4 +126,5 @@ Here is an example of how can `opt-dist` be used locally (outside of CI):
 
 > Note: if you want to run the actual CI pipeline, instead of running `opt-dist` locally,
-> you can execute `python3 src/ci/github-actions/ci.py run-local dist-x86_64-linux`.
+python3 src/ci/github-actions/ci.py run-local
```
```suggestion
```
typo

Kobzol replied:
Oops, thanks!

marcoieni commented on .github/workflows/ci.yml:
```diff
@@ -1,8 +1,8 @@
 #
-# The actual definition of the executed jobs is calculated by a Python
-# script located at src/ci/github-actions/ci.py, which
+# The actual definition of the executed jobs is calculated by a
+# script located at src/ci/citool, which
```
```suggestion
# The actual definition of the executed jobs is calculated by
# the src/ci/citool crate, which
```

marcoieni reviewed (commented):
comments on the documentation

jieyouxu commented:
There is a risk (arguably tiny) that may or may not be worth considering, which is that with a stable toolchain bump that this somehow can't build anymore (which would be a bit awkward)[^note]. Is it somehow possible to pin which rust toolchain is used to build this in extraordinary scenarios?

[^note]: for normal crates I would not have considered this, but this is the CI which kinda needs to keep working so

marcoieni commented:
I wrote a test. Add this to Cargo.toml:

```toml
[dev-dependencies]
insta = "1"
```

test file:

```rust
use std::process::{Command, Stdio};

const TEST_JOBS_YML_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/jobs.yml");

#[test]
fn test_cargo_check() {
    let output = Command::new("cargo")
        .args(["run", "-q", "calculate-job-matrix", "--jobs-file", TEST_JOBS_YML_PATH])
        .env("GITHUB_EVENT_NAME", "push")
        .env("COMMIT_MESSAGE", "test commit message")
        .env("GITHUB_REF", "refs/heads/auto")
        .stdout(Stdio::piped())
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8(output.stdout).unwrap();
    let stderr = String::from_utf8(output.stderr).unwrap();
    if !output.status.success() {
        panic!("cargo run failed: {}\n{}", stdout, stderr);
    }
    insta::assert_snapshot!(stdout, @r#"
    jobs=[{"name":"aarch64-gnu","full_name":"auto - aarch64-gnu","os":"ubuntu-22.04-arm","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","TOOLSTATE_PUBLISH":"1"},"free_disk":true},{"name":"x86_64-gnu-llvm-18-1","full_name":"auto - x86_64-gnu-llvm-18-1","os":"ubuntu-24.04","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","DOCKER_SCRIPT":"stage_2_test_set1.sh","IMAGE":"x86_64-gnu-llvm-18","READ_ONLY_SRC":"'0'","RUST_BACKTRACE":"1","TOOLSTATE_PUBLISH":"1"},"free_disk":true},{"name":"aarch64-apple","full_name":"auto - aarch64-apple","os":"macos-14","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","MACOSX_DEPLOYMENT_TARGET":"11.0","MACOSX_STD_DEPLOYMENT_TARGET":"11.0","NO_DEBUG_ASSERTIONS":"1","NO_LLVM_ASSERTIONS":"1","NO_OVERFLOW_CHECKS":"1","RUSTC_RETRY_LINKER_ON_SEGFAULT":"1","RUST_CONFIGURE_ARGS":"--enable-sanitizers --enable-profiler --set rust.jemalloc","SCRIPT":"./x.py --stage 2 test --host=aarch64-apple-darwin --target=aarch64-apple-darwin","SELECT_XCODE":"/Applications/Xcode_15.4.app","TOOLSTATE_PUBLISH":"1","USE_XCODE_CLANG":"1"}},{"name":"dist-i686-msvc","full_name":"auto - dist-i686-msvc","os":"windows-2022","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","CODEGEN_BACKENDS":"llvm,cranelift","DEPLOY_BUCKET":"rust-lang-ci2","DIST_REQUIRE_ALL_TOOLS":"1","RUST_CONFIGURE_ARGS":"--build=i686-pc-windows-msvc --host=i686-pc-windows-msvc --target=i686-pc-windows-msvc,i586-pc-windows-msvc --enable-full-tools --enable-profiler","SCRIPT":"python x.py dist bootstrap --include-default-paths","TOOLSTATE_PUBLISH":"1"}}]
    run_type=auto
    "#);
}
```

diffs of main.rs:

```diff
diff --git a/src/ci/citool/src/main.rs b/src/ci/citool/src/main.rs
index 9a22d1264fa..4c6591443a2 100644
--- a/src/ci/citool/src/main.rs
+++ b/src/ci/citool/src/main.rs
@@ -1,5 +1,5 @@
-use std::collections::HashMap;
-use std::path::Path;
+use std::collections::BTreeMap;
+use std::path::{Path, PathBuf};
 use std::process::Command;
 
 use anyhow::Context;
@@ -17,13 +17,13 @@ struct Job {
     name: String,
     /// GitHub runner on which the job should be executed
     os: String,
-    env: HashMap<String, Value>,
+    env: BTreeMap<String, Value>,
     /// Should the job be only executed on a specific channel?
     #[serde(default)]
     only_on_channel: Option<String>,
     /// Rest of attributes that will be passed through to GitHub actions
     #[serde(flatten)]
-    extra_keys: HashMap<String, Value>,
+    extra_keys: BTreeMap<String, Value>,
 }
 
 impl Job {
@@ -44,11 +44,11 @@ fn image(&self) -> String {
 #[derive(serde::Deserialize, Debug)]
 struct JobEnvironments {
     #[serde(rename = "pr")]
-    pr_env: HashMap<String, Value>,
+    pr_env: BTreeMap<String, Value>,
     #[serde(rename = "try")]
-    try_env: HashMap<String, Value>,
+    try_env: BTreeMap<String, Value>,
     #[serde(rename = "auto")]
-    auto_env: HashMap<String, Value>,
+    auto_env: BTreeMap<String, Value>,
 }
 
 #[derive(serde::Deserialize, Debug)]
@@ -71,7 +71,7 @@ fn find_auto_job_by_name(&self, name: &str) -> Option<Job> {
 }
 
 fn load_job_db(path: &Path) -> anyhow::Result<JobDatabase> {
-    let db = std::fs::read_to_string(path)?;
+    let db = read_to_string(path)?;
     let mut db: Value = serde_yaml::from_str(&db)?;
 
     // We need to expand merge keys (<<), because serde_yaml can't deal with them
@@ -92,9 +92,9 @@ struct GithubActionsJob {
     /// prefix (PR/try/auto).
     full_name: String,
     os: String,
-    env: HashMap<String, String>,
+    env: BTreeMap<String, String>,
     #[serde(flatten)]
-    extra_keys: HashMap<String, serde_json::Value>,
+    extra_keys: BTreeMap<String, serde_json::Value>,
 }
 
 /// Type of workflow that is being executed on CI
@@ -175,7 +175,7 @@ fn skip_jobs(jobs: Vec<Job>, channel: &str) -> Vec<Job> {
         .collect()
 }
 
-fn to_string_map(map: &HashMap<String, Value>) -> HashMap<String, String> {
+fn to_string_map(map: &BTreeMap<String, Value>) -> BTreeMap<String, String> {
     map.iter()
         .map(|(key, value)| {
             (
@@ -232,7 +232,7 @@ fn calculate_jobs(
     let jobs = jobs
         .into_iter()
         .map(|job| {
-            let mut env: HashMap<String, String> = to_string_map(base_env);
+            let mut env: BTreeMap<String, String> = to_string_map(base_env);
             env.extend(to_string_map(&job.env));
             let full_name = format!("{prefix} - {}", job.name);
 
@@ -314,7 +314,7 @@ fn run_workflow_locally(db: JobDatabase, job_type: JobType, name: String) -> any
     };
     let job = find_linux_job(&jobs, &name).with_context(|| format!("Cannot find job {name}"))?;
 
-    let mut custom_env: HashMap<String, String> = HashMap::new();
+    let mut custom_env: BTreeMap<String, String> = BTreeMap::new();
     // Replicate src/ci/scripts/setup-environment.sh
     // Adds custom environment variables to the job
     if name.starts_with("dist-") {
@@ -340,7 +340,10 @@ fn run_workflow_locally(db: JobDatabase, job_type: JobType, name: String) -> any
 enum Args {
     /// Calculate a list of jobs that should be executed on CI.
     /// Should only be used on CI inside GitHub actions.
-    CalculateJobMatrix,
+    CalculateJobMatrix {
+        #[clap(long)]
+        jobs_file: Option<PathBuf>,
+    },
     /// Execute a given CI job locally.
     #[clap(name = "run-local")]
     RunJobLocally {
@@ -362,19 +365,29 @@ enum JobType {
 
 fn main() -> anyhow::Result<()> {
     let args = Args::parse();
-    let db = load_job_db(Path::new(JOBS_YML_PATH)).context("Cannot load jobs.yml")?;
+    let default_jobs_file = Path::new(JOBS_YML_PATH);
+    let default_db = |jobs_path| load_job_db(jobs_path).context("Cannot load jobs.yml");
 
     match args {
-        Args::CalculateJobMatrix => {
+        Args::CalculateJobMatrix { jobs_file } => {
+            let jobs_path = jobs_file.as_deref().unwrap_or(default_jobs_file);
             let gh_ctx = load_github_ctx()
                 .context("Cannot load environment variables from GitHub Actions")?;
-            let channel = std::fs::read_to_string(Path::new(CI_DIRECTORY).join("channel"))
+            let channel = read_to_string(Path::new(CI_DIRECTORY).join("channel"))
                 .context("Cannot read channel file")?;
 
-            calculate_job_matrix(db, gh_ctx, &channel).context("Failed to calculate job matrix")?;
+            calculate_job_matrix(default_db(jobs_path)?, gh_ctx, &channel)
+                .context("Failed to calculate job matrix")?;
+        }
+        Args::RunJobLocally { job_type, name } => {
+            run_workflow_locally(default_db(default_jobs_file)?, job_type, name)?
         }
-        Args::RunJobLocally { job_type, name } => run_workflow_locally(db, job_type, name)?,
     }
 
     Ok(())
 }
+
+fn read_to_string<P: AsRef<Path>>(path: P) -> anyhow::Result<String> {
+    let error = format!("Cannot read file {:?}", path.as_ref());
+    std::fs::read_to_string(path).context(error)
+}
```

tests/jobs.yml:

```yaml
runners:
  - &base-job
    env: { }

  - &job-linux-4c
    os: ubuntu-24.04
    # Free some disk space to avoid running out of space during the build.
    free_disk: true
    <<: *base-job

  - &job-linux-16c
    os: ubuntu-22.04-16core-64gb
    <<: *base-job

  - &job-macos-m1
    os: macos-14
    <<: *base-job

  - &job-windows
    os: windows-2022
    <<: *base-job

  - &job-aarch64-linux
    # Free some disk space to avoid running out of space during the build.
    free_disk: true
    os: ubuntu-22.04-arm
    <<: *base-job
envs:
  env-x86_64-apple-tests: &env-x86_64-apple-tests
    SCRIPT: ./x.py --stage 2 test --skip tests/ui --skip tests/rustdoc -- --exact
    RUST_CONFIGURE_ARGS: --build=x86_64-apple-darwin --enable-sanitizers --enable-profiler --set rust.jemalloc
    RUSTC_RETRY_LINKER_ON_SEGFAULT: 1
    # Ensure that host tooling is tested on our minimum supported macOS version.
    MACOSX_DEPLOYMENT_TARGET: 10.12
    MACOSX_STD_DEPLOYMENT_TARGET: 10.12
    SELECT_XCODE: /Applications/Xcode_15.2.app
    NO_LLVM_ASSERTIONS: 1
    NO_DEBUG_ASSERTIONS: 1
    NO_OVERFLOW_CHECKS: 1

  production:
    &production
    DEPLOY_BUCKET: rust-lang-ci2
    # AWS_SECRET_ACCESS_KEYs are stored in GitHub's secrets storage, named
    # AWS_SECRET_ACCESS_KEY_<keyid>. Including the key id in the name allows to
    # rotate them in a single branch while keeping the old key in another
    # branch, which wouldn't be possible if the key was named with the kind
    # (caches, artifacts...).
    CACHES_AWS_ACCESS_KEY_ID: AKIA46X5W6CZI5DHEBFL
    ARTIFACTS_AWS_ACCESS_KEY_ID: AKIA46X5W6CZN24CBO55
    AWS_REGION: us-west-1
    TOOLSTATE_PUBLISH: 1

  try:
    <<: *production
    # The following env var activates faster `try` builds in `opt-dist` by, e.g.
    # - building only the more commonly useful components (we rarely need e.g. rust-docs in try
    #   builds)
    # - not running `opt-dist`'s post-optimization smoke tests on the resulting toolchain
    #
    # If you *want* these to happen however, temporarily comment it before triggering a try build.
    DIST_TRY_BUILD: 1

  auto:
    <<: *production

  pr:
    PR_CI_JOB: 1

# Jobs that run on each push to a pull request (PR)
# These jobs automatically inherit envs.pr, to avoid repeating
# it in each job definition.
pr:
  - name: mingw-check
    <<: *job-linux-4c
  - name: mingw-check-tidy
    continue_on_error: true
    <<: *job-linux-4c

# Jobs that run when you perform a try build (@bors try)
# These jobs automatically inherit envs.try, to avoid repeating
# it in each job definition.
try:
  - name: dist-x86_64-linux
    env:
      CODEGEN_BACKENDS: llvm,cranelift
    <<: *job-linux-16c

# Main CI jobs that have to be green to merge a commit into master
# These jobs automatically inherit envs.auto, to avoid repeating
# it in each job definition.
auto:
  - name: aarch64-gnu
    <<: *job-aarch64-linux

  # The x86_64-gnu-llvm-18 job is split into multiple jobs to run tests in parallel.
  # x86_64-gnu-llvm-18-1 skips tests that run in x86_64-gnu-llvm-18-{2,3}.
  - name: x86_64-gnu-llvm-18-1
    env:
      RUST_BACKTRACE: 1
      READ_ONLY_SRC: "0"
      IMAGE: x86_64-gnu-llvm-18
      DOCKER_SCRIPT: stage_2_test_set1.sh
    <<: *job-linux-4c


  ####################
  #  macOS Builders  #
  ####################

  - name: aarch64-apple
    env:
      SCRIPT: ./x.py --stage 2 test --host=aarch64-apple-darwin --target=aarch64-apple-darwin
      RUST_CONFIGURE_ARGS: >-
        --enable-sanitizers
        --enable-profiler
        --set rust.jemalloc
      RUSTC_RETRY_LINKER_ON_SEGFAULT: 1
      SELECT_XCODE: /Applications/Xcode_15.4.app
      USE_XCODE_CLANG: 1
      # Aarch64 tooling only needs to support macOS 11.0 and up as nothing else
      # supports the hardware, so only need to test it there.
      MACOSX_DEPLOYMENT_TARGET: 11.0
      MACOSX_STD_DEPLOYMENT_TARGET: 11.0
      NO_LLVM_ASSERTIONS: 1
      NO_DEBUG_ASSERTIONS: 1
      NO_OVERFLOW_CHECKS: 1
    <<: *job-macos-m1

  ######################
  #  Windows Builders  #
  ######################

  - name: dist-i686-msvc
    env:
      RUST_CONFIGURE_ARGS: >-
        --build=i686-pc-windows-msvc
        --host=i686-pc-windows-msvc
        --target=i686-pc-windows-msvc,i586-pc-windows-msvc
        --enable-full-tools
        --enable-profiler
      SCRIPT: python x.py dist bootstrap --include-default-paths
      DIST_REQUIRE_ALL_TOOLS: 1
      CODEGEN_BACKENDS: llvm,cranelift
    <<: *job-windows
```

What do you think?

Kobzol commented:
> There is a risk (arguably tiny) that may or may not be worth considering, which is that with a stable toolchain bump (through GHA's rustup I think?) that this somehow can't build anymore[1](https://github.com/rust-lang/rust/pull/136864#user-content-fn-elaboration-5c44341a6c34e93ce954d35c6d1630ca) or fail at runtime due to $reasons (which would be a bit awkward)[2](https://github.com/rust-lang/rust/pull/136864#user-content-fn-note-5c44341a6c34e93ce954d35c6d1630ca)[3](https://github.com/rust-lang/rust/pull/136864#user-content-fn-brainstorm-5c44341a6c34e93ce954d35c6d1630ca). Is it somehow possible to pin which rust toolchain is used to build this in extraordinary scenarios?

It should be possible to pin the toolchain, but it would also mean that we'd need to download a new Rust toolchain on every workflow execution. Not a big deal, but still would be nice to avoid I guess.

The same situation held also for the previous Python script, updates to the Ubuntu base image could have broken it. But we manually control which runner image we use, so unless you consider Rust to be severely less stable than Python in this regard, I think it should be fine. It will be a bit more brittle because dependencies will have to be downloaded from crates.io, which can fail sometimes, but if crates.io fails, the rest of the CI workflow would fail anyway..

marcoieni commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,380 @@
+    eprintln!("Executing {cmd:?}");
+
+    let result = cmd.spawn()?.wait()?;
+    if !result.success() { Err(anyhow::anyhow!("Job failed")) } else { Ok(()) }
```
```suggestion
    result.success().then_some(()).ok_or_else(|| anyhow::anyhow!("Job failed"))
```
choose the style you prefer 👍

marcoieni commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,380 @@
+            let is_try_build =
+                ["refs/heads/try", "refs/heads/try-perf", "refs/heads/automation/bors/try"]
+                    .iter()
+                    .any(|r| **r == self.branch_ref);
```
```suggestion
                    .any(|r| *r == self.branch_ref);
```
one is enough 👍

marcoieni commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,380 @@
+
+impl GitHubContext {
+    fn get_run_type(&self) -> Option<RunType> {
+        if self.event_name == "pull_request" {
```
this can also be a match 👍

Kobzol committed:
Add comment to [workspace]

Kobzol commented on src/ci/citool/src/main.rs:
```diff
@@ -0,0 +1,380 @@
+        JobType::Auto => &db.auto_jobs,
+        JobType::PR => &db.pr_jobs,
+    };
+    let job = find_linux_job(&jobs, &name).with_context(|| format!("Cannot find job {name}"))?;
```
I don't think it's very important.

Kobzol committed:
Add tests

Kobzol commented:
Great point with `BTreeMap`, we should strive for determinism here. And also a good idea with the tests, they already helped me find a bug :laughing:

jieyouxu commented:
> The same situation held also for the previous Python script, updates to the Ubuntu base image could have broken it. But we manually control which runner image we use, so unless you consider Rust to be severely less stable than Python in this regard, I think it should be fine. It will be a bit more brittle because dependencies will have to be downloaded from crates.io, which can fail sometimes, but if crates.io fails, the rest of the CI workflow would fail anyway..

Of course, I'm not saying the python version didn't have this issue :P

marcoieni commented:
LGTM. You can r=me when you feel it's ready 👍
It might be nice to check if the code is formatted and tests pass in CI.

Kobzol commented:
Thanks! I added running tests on CI. I'll do a few tests with try builds.

@bors try

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13267767235/job/37039753770)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13267767235/job/37039753770","builder_name":"checks-actions"} -->

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit 39d61bc37ec54f5846525be4239e7d0c8eec8995 with merge 372fac420652156e9a5b3343a4f5633399909771...
<!-- homu: {"type":"TryBuildStarted","head_sha":"39d61bc37ec54f5846525be4239e7d0c8eec8995","merge_sha":"372fac420652156e9a5b3343a4f5633399909771"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13267879131/job/37040107035)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13267879131/job/37040107035","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13267718556/job/37039336844) [(plain)](https://github.com/rust-lang/rust/commit/39d61bc37ec54f5846525be4239e7d0c8eec8995/checks/37039336844/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.804 Building wheels for collected packages: reuse
#16 2.805   Building wheel for reuse (pyproject.toml): started
#16 3.017   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.018   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=be6760d5849de4a58bbe52b85ca57a55f2b32b518b17029a5ad2e530db0d4303
#16 3.018   Stored in directory: /tmp/pip-ephem-wheel-cache-40w3dm3s/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.021 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.413 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.413 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
```

</details>

Kobzol commented:
Uhh, this is a bit annoying, the CI script uninstall rustup (and stable cargo and rustc) at its beginning, so it's not at all trivial to run the tests, and I don't feel like adding this to bootstrap.

Kobzol committed:
Test citool on CI

jieyouxu commented:
Apparently https://github.com/rust-lang/rust/issues/110479 exists, heh

Kobzol commented:
@bors r=marcoieni rollup=never

bors commented:
:pushpin: Commit 591bbbcb91bee3ebe19210991010b628637ec6b9 has been approved by `marcoieni`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=marcoieni 591bbbcb91bee3ebe19210991010b628637ec6b9 -->
<!-- homu: {"type":"Approved","sha":"591bbbcb91bee3ebe19210991010b628637ec6b9","approver":"marcoieni","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
Rebased.

@bors r=marcoieni

bors commented:
:pushpin: Commit 1f521956c4b79e85f27f81e6ccd8d215b325ad04 has been approved by `marcoieni`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=marcoieni 1f521956c4b79e85f27f81e6ccd8d215b325ad04 -->
<!-- homu: {"type":"Approved","sha":"1f521956c4b79e85f27f81e6ccd8d215b325ad04","approver":"marcoieni","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 1f521956c4b79e85f27f81e6ccd8d215b325ad04 with merge cc40d6f1d15b018d63a7848c3bae6b1f2d015a99...
<!-- homu: {"type":"BuildStarted","head_sha":"1f521956c4b79e85f27f81e6ccd8d215b325ad04","merge_sha":"cc40d6f1d15b018d63a7848c3bae6b1f2d015a99"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13336857215/job/37253975612)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13336857215/job/37253975612","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13336857215/job/37253932459) [(plain)](https://github.com/rust-lang-ci/rust/commit/cc40d6f1d15b018d63a7848c3bae6b1f2d015a99/checks/37253932459/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
/dev/sdb1       147G  4.1G  136G   3% /mnt
tmpfs           1.6G  4.0K  1.6G   1% /run/user/1001
================================================================================

Failed to stop mysql.service: Unit mysql.service not loaded.
##[error]Process completed with exit code 5.
```

</details>

Kobzol commented:
@bors retry mysql error

bors commented:
:hourglass: Testing commit 1f521956c4b79e85f27f81e6ccd8d215b325ad04 with merge 996a9df0e9d85d8647a2e169831d83af2ed41c12...
<!-- homu: {"type":"BuildStarted","head_sha":"1f521956c4b79e85f27f81e6ccd8d215b325ad04","merge_sha":"996a9df0e9d85d8647a2e169831d83af2ed41c12"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13356185704/job/37299328596)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13356185704/job/37299328596","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13356185704/job/37299317459) [(plain)](https://github.com/rust-lang-ci/rust/commit/996a9df0e9d85d8647a2e169831d83af2ed41c12/checks/37299317459/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
adding extra environment variable DEPLOY_BUCKET
adding extra environment variable RUST_CONFIGURE_ARGS
adding extra environment variable SCRIPT
adding extra environment variable TOOLSTATE_PUBLISH
##[error]Unable to process file command 'env' successfully.
##[error]Invalid format '  python x.py test --stage 2 src/tools/miri --target aarch64-apple-darwin --test-args pass && python x.py test --stage 2 src/tools/miri --target i686-pc-windows-gnu --test-args pass && python x.py miri --stage 2 library/core --test-args notest && python x.py miri --stage 2 library/alloc --test-args notest && python x.py miri --stage 2 library/std --test-args notest'
[command]"C:\Program Files\Git\bin\git.exe" version
git version 2.47.1.windows.2
Copying 'C:\Users\runneradmin\.gitconfig' to 'D:\a\_temp\20065385-16d6-4433-b989-bb4bc16843ec\.gitconfig'
Temporarily overriding HOME='D:\a\_temp\20065385-16d6-4433-b989-bb4bc16843ec' before making global git config changes
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #136914) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Kobzol commented:
Hmm. I will run the Rust and Python versions locally and make sure that they return the exact same output.

Kobzol committed:
Avoid double serialization of environment strings

Kobzol committed:
Trim try-job names

Kobzol committed:
Trim channel

To avoid including a newline at the end, which broke `only_on_channel` comparison.

Kobzol commented:
Fixed handling of environment variables and reading of the `src/ci/channel` file. I tested locally that the Rust version should now return the exact same output as the Python version for PR/try/auto builds.

@bors try

bors commented:
:hourglass: Trying commit 6a41c3f58c5682482408ce0b85f207b0e691f5af with merge 29559ee0b67de151c068118494be122cd89806ee...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6a41c3f58c5682482408ce0b85f207b0e691f5af","merge_sha":"29559ee0b67de151c068118494be122cd89806ee"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13369227726/job/37333835886) [(plain)](https://github.com/rust-lang-ci/rust/commit/29559ee0b67de151c068118494be122cd89806ee/checks/37333835886/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- pr_jobs stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: pr_jobs
Source: tests/jobs.rs:42
Expression: stdout
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
+new results
────────────┬───────────────────────────────────────────────────────────────────
    0       │-jobs=[{"name":"mingw-check","full_name":"PR - mingw-check","os":"ubuntu-24.04","env":{"PR_CI_JOB":"1"},"free_disk":true},{"name":"mingw-check-tidy","full_name":"PR - mingw-check-tidy","os":"ubuntu-24.04","env":{"PR_CI_JOB":"1"},"continue_on_error":true,"free_disk":true}]
          0 │+jobs=[{"name":"mingw-check","full_name":"PR - mingw-check","os":"ubuntu-24.04","env":{"PR_CI_JOB":1},"free_disk":true},{"name":"mingw-check-tidy","full_name":"PR - mingw-check-tidy","os":"ubuntu-24.04","env":{"PR_CI_JOB":1},"continue_on_error":true,"free_disk":true}]
    1     1 │ run_type=pr
error: test failed, to rerun pass `--test jobs`
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'pr_jobs' panicked at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.42.1/src/runtime.rs:679:13:
snapshot assertion for 'pr_jobs' failed in line 42

---- try_jobs stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: try_jobs
Snapshot: try_jobs
Source: tests/jobs.rs:17
────────────────────────────────────────────────────────────────────────────────
Expression: stdout
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
    0       │-jobs=[{"name":"dist-x86_64-linux","full_name":"try - dist-x86_64-linux","os":"ubuntu-22.04-16core-64gb","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","CODEGEN_BACKENDS":"llvm,cranelift","DEPLOY_BUCKET":"rust-lang-ci2","DIST_TRY_BUILD":"1","TOOLSTATE_PUBLISH":"1"}}]
          0 │+jobs=[{"name":"dist-x86_64-linux","full_name":"try - dist-x86_64-linux","os":"ubuntu-22.04-16core-64gb","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","CODEGEN_BACKENDS":"llvm,cranelift","DEPLOY_BUCKET":"rust-lang-ci2","DIST_TRY_BUILD":1,"TOOLSTATE_PUBLISH":1}}]
    1     1 │ run_type=try
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'try_jobs' panicked at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.42.1/src/runtime.rs:679:13:
snapshot assertion for 'try_jobs' failed in line 17

---- try_custom_jobs stdout ----
Snapshot: try_custom_jobs
Source: tests/jobs.rs:33
────────────────────────────────────────────────────────────────────────────────
Expression: stdout
Expression: stdout
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
    0       │-jobs=[{"name":"aarch64-gnu","full_name":"try - aarch64-gnu","os":"ubuntu-22.04-arm","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","DIST_TRY_BUILD":"1","TOOLSTATE_PUBLISH":"1"},"free_disk":true},{"name":"dist-i686-msvc","full_name":"try - dist-i686-msvc","os":"windows-2022","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","CODEGEN_BACKENDS":"llvm,cranelift","DEPLOY_BUCKET":"rust-lang-ci2","DIST_REQUIRE_ALL_TOOLS":"1","DIST_TRY_BUILD":"1","RUST_CONFIGURE_ARGS":"--build=i686-pc-windows-msvc --host=i686-pc-windows-msvc --target=i686-pc-windows-msvc,i586-pc-windows-msvc --enable-full-tools --enable-profiler","SCRIPT":"python x.py dist bootstrap --include-default-paths","TOOLSTATE_PUBLISH":"1"}}]
          0 │+jobs=[{"name":"aarch64-gnu","full_name":"try - aarch64-gnu","os":"ubuntu-22.04-arm","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","DIST_TRY_BUILD":1,"TOOLSTATE_PUBLISH":1},"free_disk":true},{"name":"dist-i686-msvc","full_name":"try - dist-i686-msvc","os":"windows-2022","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","CODEGEN_BACKENDS":"llvm,cranelift","DEPLOY_BUCKET":"rust-lang-ci2","DIST_REQUIRE_ALL_TOOLS":1,"DIST_TRY_BUILD":1,"RUST_CONFIGURE_ARGS":"--build=i686-pc-windows-msvc --host=i686-pc-windows-msvc --target=i686-pc-windows-msvc,i586-pc-windows-msvc --enable-full-tools --enable-profiler","SCRIPT":"python x.py dist bootstrap --include-default-paths","TOOLSTATE_PUBLISH":1}}]
    1     1 │ run_type=try
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'try_custom_jobs' panicked at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.42.1/src/runtime.rs:679:13:
snapshot assertion for 'try_custom_jobs' failed in line 33

---- auto_jobs stdout ----
Snapshot: auto_jobs
Source: tests/jobs.rs:8
────────────────────────────────────────────────────────────────────────────────
Expression: stdout
Expression: stdout
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
    0       │-jobs=[{"name":"aarch64-gnu","full_name":"auto - aarch64-gnu","os":"ubuntu-22.04-arm","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","TOOLSTATE_PUBLISH":"1"},"free_disk":true},{"name":"x86_64-gnu-llvm-18-1","full_name":"auto - x86_64-gnu-llvm-18-1","os":"ubuntu-24.04","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","DOCKER_SCRIPT":"stage_2_test_set1.sh","IMAGE":"x86_64-gnu-llvm-18","READ_ONLY_SRC":"'0'","RUST_BACKTRACE":"1","TOOLSTATE_PUBLISH":"1"},"free_disk":true},{"name":"aarch64-apple","full_name":"auto - aarch64-apple","os":"macos-14","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","MACOSX_DEPLOYMENT_TARGET":"11.0","MACOSX_STD_DEPLOYMENT_TARGET":"11.0","NO_DEBUG_ASSERTIONS":"1","NO_LLVM_ASSERTIONS":"1","NO_OVERFLOW_CHECKS":"1","RUSTC_RETRY_LINKER_ON_SEGFAULT":"1","RUST_CONFIGURE_ARGS":"--enable-sanitizers --enable-profiler --set rust.jemalloc","SCRIPT":"./x.py --stage 2 test --host=aarch64-apple-darwin --target=aarch64-apple-darwin","SELECT_XCODE":"/Applications/Xcode_15.4.app","TOOLSTATE_PUBLISH":"1","USE_XCODE_CLANG":"1"}},{"name":"dist-i686-msvc","full_name":"auto - dist-i686-msvc","os":"windows-2022","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","CODEGEN_BACKENDS":"llvm,cranelift","DEPLOY_BUCKET":"rust-lang-ci2","DIST_REQUIRE_ALL_TOOLS":"1","RUST_CONFIGURE_ARGS":"--build=i686-pc-windows-msvc --host=i686-pc-windows-msvc --target=i686-pc-windows-msvc,i586-pc-windows-msvc --enable-full-tools --enable-profiler","SCRIPT":"python x.py dist bootstrap --include-default-paths","TOOLSTATE_PUBLISH":"1"}}]
          0 │+jobs=[{"name":"aarch64-gnu","full_name":"auto - aarch64-gnu","os":"ubuntu-22.04-arm","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","TOOLSTATE_PUBLISH":1},"free_disk":true},{"name":"x86_64-gnu-llvm-18-1","full_name":"auto - x86_64-gnu-llvm-18-1","os":"ubuntu-24.04","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","DOCKER_SCRIPT":"stage_2_test_set1.sh","IMAGE":"x86_64-gnu-llvm-18","READ_ONLY_SRC":"0","RUST_BACKTRACE":1,"TOOLSTATE_PUBLISH":1},"free_disk":true},{"name":"aarch64-apple","full_name":"auto - aarch64-apple","os":"macos-14","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","DEPLOY_BUCKET":"rust-lang-ci2","MACOSX_DEPLOYMENT_TARGET":11.0,"MACOSX_STD_DEPLOYMENT_TARGET":11.0,"NO_DEBUG_ASSERTIONS":1,"NO_LLVM_ASSERTIONS":1,"NO_OVERFLOW_CHECKS":1,"RUSTC_RETRY_LINKER_ON_SEGFAULT":1,"RUST_CONFIGURE_ARGS":"--enable-sanitizers --enable-profiler --set rust.jemalloc","SCRIPT":"./x.py --stage 2 test --host=aarch64-apple-darwin --target=aarch64-apple-darwin","SELECT_XCODE":"/Applications/Xcode_15.4.app","TOOLSTATE_PUBLISH":1,"USE_XCODE_CLANG":1}},{"name":"dist-i686-msvc","full_name":"auto - dist-i686-msvc","os":"windows-2022","env":{"ARTIFACTS_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZN24CBO55","AWS_REGION":"us-west-1","CACHES_AWS_ACCESS_KEY_ID":"AKIA46X5W6CZI5DHEBFL","CODEGEN_BACKENDS":"llvm,cranelift","DEPLOY_BUCKET":"rust-lang-ci2","DIST_REQUIRE_ALL_TOOLS":1,"RUST_CONFIGURE_ARGS":"--build=i686-pc-windows-msvc --host=i686-pc-windows-msvc --target=i686-pc-windows-msvc,i586-pc-windows-msvc --enable-full-tools --enable-profiler","SCRIPT":"python x.py dist bootstrap --include-default-paths","TOOLSTATE_PUBLISH":1}}]
    1     1 │ run_type=auto
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'auto_jobs' panicked at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.42.1/src/runtime.rs:679:13:
snapshot assertion for 'auto_jobs' failed in line 8

failures:
    auto_jobs
    pr_jobs
```

</details>

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit 31acbd3f902da4eba6fe7eaf07cb862ebac9d67c with merge 457e90245a010094d720fde6243f234e9d25e103...
<!-- homu: {"type":"TryBuildStarted","head_sha":"31acbd3f902da4eba6fe7eaf07cb862ebac9d67c","merge_sha":"457e90245a010094d720fde6243f234e9d25e103"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13369276395/job/37345237776)
Build commit: 457e90245a010094d720fde6243f234e9d25e103 (`457e90245a010094d720fde6243f234e9d25e103`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13369276395/job/37345237776"},"merge_sha":"457e90245a010094d720fde6243f234e9d25e103"} -->

Kobzol commented:
Ok now I'm confident that the script works as it should.

@rustbot ready

Kobzol commented:
@marcoieni Once you have a bit of time, could you please recheck if the latest changes look fine, and if so, reapprove? I didn't want to r=you without you taking another look.

I have like five branches stacked on top of this :joy: So wanted to get it through the queue. Thanks!

marcoieni commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit 31acbd3f902da4eba6fe7eaf07cb862ebac9d67c has been approved by `marcoieni`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=marcoieni 31acbd3f902da4eba6fe7eaf07cb862ebac9d67c -->
<!-- homu: {"type":"Approved","sha":"31acbd3f902da4eba6fe7eaf07cb862ebac9d67c","approver":"marcoieni","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors p=4 (scheduling)

bors commented:
:hourglass: Testing commit 31acbd3f902da4eba6fe7eaf07cb862ebac9d67c with merge 4b696e6bf77b4df02634d346637bd35f2960c5f4...
<!-- homu: {"type":"BuildStarted","head_sha":"31acbd3f902da4eba6fe7eaf07cb862ebac9d67c","merge_sha":"4b696e6bf77b4df02634d346637bd35f2960c5f4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13614030973/job/38057935537)
Approved by: marcoieni
Pushing 4b696e6bf77b4df02634d346637bd35f2960c5f4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"marcoieni","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13614030973/job/38057935537"},"merge_sha":"4b696e6bf77b4df02634d346637bd35f2960c5f4"} -->

rust-timer commented:
Finished benchmarking commit (4b696e6bf77b4df02634d346637bd35f2960c5f4): [comparison URL](https://perf.rust-lang.org/compare.html?start=1c3b035542775e9a5decc93167d351b062942d32&end=4b696e6bf77b4df02634d346637bd35f2960c5f4&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1c3b035542775e9a5decc93167d351b062942d32&end=4b696e6bf77b4df02634d346637bd35f2960c5f4&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1c3b035542775e9a5decc93167d351b062942d32&end=4b696e6bf77b4df02634d346637bd35f2960c5f4&stat=max-rss)
<details>
<summary>Results (primary -1.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.7% | [-1.7%, -1.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.7% | [-1.7%, -1.7%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1c3b035542775e9a5decc93167d351b062942d32&end=4b696e6bf77b4df02634d346637bd35f2960c5f4&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=1c3b035542775e9a5decc93167d351b062942d32&end=4b696e6bf77b4df02634d346637bd35f2960c5f4&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 773.253s -> 774.29s (0.13%)
**Artifact size**: 361.91 MiB -> 361.94 MiB (0.01%)
